### PR TITLE
Rule S112 - System reserved exceptions should not be thrown

### PR DESCRIPTION
**Pull request by valhristov**

### CODE REVIEW

File: src/SonarAnalyzer.Utilities/Rules.Description/S112.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<p>Throwing such generic exceptions as <code>IndexOutOfRangeException</code>, <code>NullReferenceException</code>, <code>OutOfMemoryException</code>
```
Comment (tvsonar) [inline src/SonarAnalyzer.Utilities/Rules.Description/S112.html]: There are missing exceptions here, for example "System.SystemException" is not listed

File: src/SonarAnalyzer.CSharp/Rules/ThrowReservedExceptions.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public class ThrowReservedExceptions : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S112";
+        internal const string MessageFormat = "{0} should not be thrown by user code.";
```
Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/ThrowReservedExceptions.cs]: As discussed, the exception name should be between quotes.

File: src/SonarAnalyzer.CSharp/Rules/ThrowReservedExceptions.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        private static readonly ISet<string> ReservedExceptionTypeNames =
+            ImmutableHashSet.Create(new[]
```
Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/ThrowReservedExceptions.cs]: `Create` takes `params`, you can use that too.

**Comment by valhristov** [comment]
Added a VB.NET version of the rule, fixed all suggestions

**Pull request closed without merging**